File: OneCodeTeam/How-to-use-Windows-Azure-ACS-to-authenticate-in-31212/[VB]-How to use Windows Azure ACS to authenticate in WPF application/VB/VBAzureACSAuthInWPF/SecurityModule/SWTModule.vb
Imports System.Web

Class SWTModule
    Implements IHttpModule
    'USE CONFIGURATION FILE, WEB.CONFIG, TO MANAGE THIS DATA
    Private serviceNamespace As String = "your namespace"
    'Access Control Namespaces
    Private acsHostName As String = "accesscontrol.windows.net"
    Private trustedTokenPolicyKey As String = "your signing key"
    'You generated this key in read me Running the Sample step 1.7.
    Private trustedAudience As String = "your realm"
    'In this example it should be http://localhost:12526/RESTUserService.svc

    Private Sub IHttpModule_Dispose() Implements IHttpModule.Dispose

    End Sub

    Private Sub IHttpModule_Init(context As HttpApplication) Implements IHttpModule.Init
        AddHandler context.BeginRequest, AddressOf context_BeginRequest

    End Sub

    Private Sub context_BeginRequest(sender As Object, e As EventArgs)
        ' HANDLE SWT TOKEN VALIDATION
        ' get the authorization header
        Dim headerValue As String = HttpContext.Current.Request.Headers.Get("Authorization")

        ' Check that a value is there
        If String.IsNullOrEmpty(headerValue) Then
            Throw New ApplicationException("unauthorized")
        End If

        ' Check that it starts with 'WRAP'
        If Not headerValue.StartsWith("WRAP ") Then
            Throw New ApplicationException("unauthorized")
        End If

        Dim nameValuePair As String() = headerValue.Substring("WRAP ".Length).Split(New Char() {"="c}, 2)

        If nameValuePair.Length <> 2 OrElse nameValuePair(0) <> "access_token" OrElse Not nameValuePair(1).StartsWith("\") OrElse Not nameValuePair(1).EndsWith("\") Then
            Throw New ApplicationException("unauthorized")
        End If

        ' Trim off the leading and trailing double-quotes
        Dim token As String = nameValuePair(1).Substring(1, nameValuePair(1).Length - 2)

        ' Create a token validator
        Dim validator As New TokenValidator(Me.acsHostName, Me.serviceNamespace, Me.trustedAudience, Me.trustedTokenPolicyKey)

        ' Validate the token
        If Not validator.Validate(token) Then
            Throw New ApplicationException("unauthorized")
        End If

    End Sub
End Class

File: OneCodeTeam/How-to-use-Windows-Azure-ACS-to-authenticate-in-25338/[C#]-How to use Windows Azure ACS to authenticate in WPF application/C#/CSAzureACSAuthInWPF/SecurityModule/SWTModule.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Web;

namespace SecurityModule
{
    class SWTModule : IHttpModule
    {
        //USE CONFIGURATION FILE, WEB.CONFIG, TO MANAGE THIS DATA
        string serviceNamespace = "your namespace"; //Access Control Namespaces
        string acsHostName = "accesscontrol.windows.net";
        string trustedTokenPolicyKey = "your signing key";//You generated this key in read me Running the Sample step 1.7.
        string trustedAudience = "your realm";//In this example it should be http://localhost:12526/RESTUserService.svc


        void IHttpModule.Dispose()
        {

        }

        void IHttpModule.Init(HttpApplication context)
        {
            context.BeginRequest += new EventHandler(context_BeginRequest);
        }

        void context_BeginRequest(object sender, EventArgs e)
        {
            // HANDLE SWT TOKEN VALIDATION
            // get the authorization header
            string headerValue = HttpContext.Current.Request.Headers.Get("Authorization");

            // Check that a value is there
            if (string.IsNullOrEmpty(headerValue))
            {
                throw new ApplicationException("unauthorized");
            }

            // Check that it starts with 'WRAP'
            if (!headerValue.StartsWith("WRAP "))
            {
                throw new ApplicationException("unauthorized");
            }

            string[] nameValuePair = headerValue.Substring("WRAP ".Length).Split(new char[] { '=' }, 2);

            if (nameValuePair.Length != 2 ||
                nameValuePair[0] != "access_token" ||
                !nameValuePair[1].StartsWith("\"") ||
                !nameValuePair[1].EndsWith("\""))
            {
                throw new ApplicationException("unauthorized");
            }

            // Trim off the leading and trailing double-quotes
            string token = nameValuePair[1].Substring(1, nameValuePair[1].Length - 2);

            // Create a token validator
            TokenValidator validator = new TokenValidator(
                this.acsHostName,
                this.serviceNamespace,
                this.trustedAudience,
                this.trustedTokenPolicyKey);

            // Validate the token
            if (!validator.Validate(token))
            {
                throw new ApplicationException("unauthorized");
            }

        }
    }
}

